### PR TITLE
[COS-953] Gross Retention Rate

### PR DESCRIPTION
**Pull request by grokkos**
## Changes

What types of changes does your code introduce?  _Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Build related changes
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Other (please describe below)

## Summary by CodeRabbit

- **New Features**
    - Enhanced dashboard metrics for Gross Revenue Retention, including percentage display and year-over-year data.
    - Added a search bar to the main interface for improved user navigation.

- **Enhancements**
    - Improved tooltip display logic for better clarity on contract data.
    - Enhanced error messaging for gross revenue retention data retrieval failures.

- **Documentation**
    - Updated GraphQL schema documentation to include new fields and types.

- **Tests**
    - Added integration tests for gross revenue retention data fetching.

- **Bug Fixes**
    - Fixed an issue where "No data yet" was not displayed correctly in the absence of contracts.

### CODE REVIEW

File: packages/server/customer-os-api/repository/dashboard_repository.go [hunk 1]
```diff
@@ -1516,3 +1517,121 @@ func getCorrectValueType(valueToExtract any) float64 {
...
+							 nanosecond: 0o00000000
+						 }) as beginOfMonth,
+     currentDate + duration({months: 1}) - duration({nanoseconds: 1}) as endOfMonth,
+     currentDate + duration({months: 1}) AS startOfNextMonth
+WITH currentDate.year AS year, currentDate.month AS month, beginOfMonth, endOfMonth, startOfNextMonth
+					OPTIONAL MATCH (t:Tenant {name: $tenant})<-[:CONTRACT_BELONGS_TO_TENANT]-(c:Contract_%s)-[:HAS_SERVICE]->(sli:ServiceLineItem_%s)
+WITH year, month, beginOfMonth, endOfMonth, c.id as cid, sli.parentId as sp, sli
+ORDER BY sli.startedAt ASC
+WHERE c.serviceStartedAt IS NOT NULL 
+  AND sli.startedAt < beginOfMonth
```
Comment (edifirut) [inline packages/server/customer-os-api/repository/dashboard_repository.go]: AND sli.startedAt < beginOfMonth AND (sli.endedAt IS NULL OR sli.endedAt > beginOfMonth)
should be
sli.startedAt < startOfNextMonth AND (sli.endedAt IS NULL OR sli.endedAt >= startOfNextMonth)

File: packages/server/customer-os-api/repository/dashboard_repository.go [hunk 2]
```diff
@@ -1516,3 +1517,121 @@ func getCorrectValueType(valueToExtract any) float64 {
...
+						CASE WHEN sliItem.sliStartedAt < beginOfMonth - duration({months: 3}) AND sliItem.sliEndedAt > beginOfMonth - duration({months: 4}) - duration({nanoseconds: 1}) THEN sliItem.amount ELSE 0 END,
+						CASE WHEN sliItem.sliStartedAt < beginOfMonth - duration({months: 2}) AND sliItem.sliEndedAt > beginOfMonth - duration({months: 3}) - duration({nanoseconds: 1}) THEN sliItem.amount ELSE 0 END,
+						CASE WHEN sliItem.sliStartedAt < beginOfMonth - duration({months: 1}) AND sliItem.sliEndedAt > beginOfMonth - duration({months: 2}) - duration({nanoseconds: 1}) THEN sliItem.amount ELSE 0 END,
+						CASE WHEN sliItem.sliStartedAt < beginOfMonth AND sliItem.sliEndedAt > beginOfMonth + duration({months: 1}) - duration({nanoseconds: 1}) THEN sliItem.amount ELSE 0 END
+     ]) as sliActivePerContract,
+     REDUCE(s = [], sliItem in sliPerContract | 
+         CASE WHEN sliItem.sliStartedAt < beginOfMonth AND sliItem.sliEndedAt > beginOfMonth + duration({months: 1}) - duration({nanoseconds: 1}) THEN s + sliItem ELSE s END
```
Comment (edifirut) [inline packages/server/customer-os-api/repository/dashboard_repository.go]: sliItem.sliStartedAt < beginOfMonth AND sliItem.sliEndedAt > beginOfMonth + duration({months: 1}) - duration({nanoseconds: 1})

should be

sliItem.sliStartedAt < startOfNextMonth and ( sliItem.endedAt IS NULL or sliItem.endedAt >= startOfNextMonth )

File: packages/server/customer-os-api/repository/dashboard_repository.go [hunk 3]
```diff
@@ -1516,3 +1517,121 @@ func getCorrectValueType(valueToExtract any) float64 {
...
+     ) as sliActiveInCurrentMonth
+WITH year, month, beginOfMonth, endOfMonth, cid, sliActivePerContract, sliActiveInCurrentMonth, 
+REDUCE(acc = null, i IN RANGE(0, 11) |
```
Comment (edifirut) [inline packages/server/customer-os-api/repository/dashboard_repository.go]: WITH year, month, beginOfMonth, endOfMonth, cid, sliActivePerContract, sliActiveInCurrentMonth,, REDUCE(sumArray = [], i IN RANGE(0, 11) |
    sumArray = sumArray[i] + sliActivePerContract[i]
) AS sumPerMonth

WITH year, month, beginOfMonth, endOfMonth, cid, sliActivePerContract, sliActiveInCurrentMonth,
REDUCE(acc = null, i IN RANGE(0, 11) |
    CASE WHEN sumPerMonth[i] > 0 AND acc IS NULL THEN sumPerMonth[i] ELSE acc END
) AS baseline

File: packages/server/customer-os-api/repository/dashboard_repository.go [hunk 4]
```diff
@@ -1528,3 +1529,121 @@ func getCorrectValueType(valueToExtract any) float64 {
...
+         CASE WHEN sliItem.sliStartedAt < startOfNextMonth  AND (sliItem.endedAt IS NULL or sliItem.endedAt > startOfNextMonth) THEN s + sliItem ELSE s END
+     ) as sliActiveInCurrentMonth
+WITH year, month, beginOfMonth, endOfMonth, cid, sliActivePerContract, sliActiveInCurrentMonth, 
+REDUCE(acc = null, i IN RANGE(0, 11) |
+    CASE WHEN sliActivePerContract[i] > 0 AND acc IS NULL THEN sliActivePerContract[i] ELSE acc END
+) AS baseline, 
+REDUCE(acc = 0, i IN sliActiveInCurrentMonth | acc + i.amount ) AS currentAmount
+WITH year, month, cid, baseline, currentAmount, sliActiveInCurrentMonth
+with year, month, SUM(baseline) as baselineValue, SUM(currentAmount) as currentValue
+return year, month, CASE WHEN baselineValue <> 0 THEN (currentValue / baselineValue) * 100 ELSE null END AS grossRevenueRetentionRate`, "% 12 + 1", tenant, tenant),
```
Comment (grokkos) [inline packages/server/customer-os-api/repository/dashboard_repository.go]: @edifirut  I ve tested a lot and I found where the pain point is, but havent yet find a way to overcome it. 
`sliActiveInCurrentMonth` and `currentValue` seem working fine from the debug I did. 
The part that is not working as expected is the `baseline` on line 1612.
If you have any ideas on how to do it differently? 

`WITH year, month, beginOfMonth, endOfMonth, cid, sliActivePerContract, sliActiveInCurrentMonth, 
REDUCE(acc = null, i IN RANGE(0, 11) |
    CASE WHEN sliActivePerContract[i] > 0 AND acc IS NULL THEN sliActivePerContract[i] ELSE acc END
) AS baseline, `

Comment (edifirut) [reply packages/server/customer-os-api/repository/dashboard_repository.go]: my suggestion would be to grab just this block and rework it in neo4j console. just build an array with arrays with values

WITH [
    [0,0,0,5,0,0,0,0,0,0,0,0],
    [0,2,0,0,0,0,0,0,0,0,0,0]
] as sliActivePerContract

MATCH (t:Tenant{name:"openlineai"})

return REDUCE(acc = null, i IN RANGE(0, 11) | CASE WHEN sliActivePerContract[i] > 0 AND acc IS NULL THEN sliActivePerContract[i] ELSE acc END ) AS baseline

Comment (grokkos) [reply packages/server/customer-os-api/repository/dashboard_repository.go]: cool I ve worked it on the neo4j console. 

that is a working query for the above block 

`WITH [
  [0, 0, 0, 5, 0, 0, 0, 0, 0, 1, 0, 0],
  [0, 0, 0, 0, 0, 0, 29, 0, 0, 0, 0, 0],
  [1, 0, 0, 0, 0, 0, 29, 0, 0, 0, 0, 0]
] AS sliActivePerContract

WITH REDUCE(acc = 0, i IN RANGE(0, SIZE(sliActivePerContract) - 1) |
  acc + COALESCE(HEAD([value IN sliActivePerContract[i] WHERE value <> 0 | value]), 0)
) AS baseline

RETURN baseline;`

will try to add that to the repository now

**Review by edifirut — APPROVED** [review]

**Pull request merged**